python/semantic_kernel/kernel.py

```diff
@@ -408,51 +297,121 @@ async def invoke_prompt(
             function_result=function_result,
             exception=exception or function_result.metadata.get("exception", None) if function_result else None,
         )
-        if self.function_invoked_handlers:
-            for handler in self.function_invoked_handlers.values():
-                handler(self, args)
-        return args
-
-    def on_function_invoking(
-        self, kernel_function_metadata: KernelFunctionMetadata, arguments: KernelArguments
-    ) -> FunctionInvokingEventArgs:
-        args = FunctionInvokingEventArgs(kernel_function_metadata=kernel_function_metadata, arguments=arguments)
-        if self.function_invoking_handlers:
-            for handler in self.function_invoking_handlers.values():
-                handler(self, args)
-        return args
-
-    def add_function_invoking_handler(
-        self, handler: Callable[["Kernel", FunctionInvokingEventArgs], FunctionInvokingEventArgs]
-    ) -> None:
-        self.function_invoking_handlers[id(handler)] = handler
+        for hook in self.hooks:
+            try:
+                if isinstance(hook, HOOK_PROTOCOLS["post_function_invoke"]):
+                    logger.debug(f"Running Post Function Invoke Hook: {hook.__class__.__name__}")
+                    result = hook.post_function_invoke(context=context)
+                    if inspect.isawaitable(result):
+                        await result
+            except Exception as exc:
+                logger.error(
+                    f"An error occurred in the on_function_invoked event of hook: {hook.__class__.__name__}. "
+                    f"Error description: {str(exc)}"
+                )
+            if context.is_cancel_requested:
+                raise OperationCancelledException(
+                    f"Execution was cancelled on function invoking event of pipeline step "
+                    f"{metadata.get('pipeline_step', '') if metadata else ''}: "
+                    f"{kernel_function_metadata.fully_qualified_name}."
+                )
+            if context.updated_arguments:
+                logger.info(
+                    f"Arguments updated by function_invoked_handler in pipeline step: "
+                    f"{metadata.get('pipeline_step', '') if metadata else ''}, new arguments: {context.arguments}"
+                )
+            if context.is_repeat_requested:
+                logger.info(
+                    f"Execution was repeated on function invoked event of pipeline step "
+                    f"{metadata.get('pipeline_step', '') if metadata else ''}: "
+                    f"{kernel_function_metadata.fully_qualified_name}."
+                )
 
-    def add_function_invoked_handler(
-        self, handler: Callable[["Kernel", FunctionInvokedEventArgs], FunctionInvokedEventArgs]
-    ) -> None:
-        self.function_invoked_handlers[id(handler)] = handler
+        return context
+
+    def add_hook(self, hook: object, priority: int | None = None) -> None:
```

Comment:
Not sure if `priority` parameter should be part of public API. It's still should be possible to configure filter ordering, like `kernel.hooks.insert(1, hook)`. Here is an example of public API in .NET:
https://github.com/microsoft/semantic-kernel/blob/0de7d34be3de11ffc86f9548b82ef0e888c27ea9/dotnet/src/SemanticKernel.UnitTests/Filters/KernelFilterTests.cs#L573-L575

If you will decide to keep this parameter, probably it would be better to rename it to `position` to identify that this is hook order of execution.

Reply:
changed to position, good call.

Reply:
but we do not want someone to add directly to hooks, now changed to _hooks to show that, since checks are done on adding.